RavenDB-20954 - SelectedNodeTag operations reach a different node when node is in rehab or when topology updates are disabled

lastav5 opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-20954

### Additional description

- operations with `SelectedNodeTag` (regardless of `disableTopologyUpdate`) did a failover to a preferred node which wasn't the *selected* node (thus we were succeeding when we expected to fail). Fix was: calling`ChooseNodeForRequest` on retry instead of `GetPreferredNode`
- operations that are waited on (patch etc) did not go to the correct node where the operation resides because `disableTopologyUpdate` skipped their `SelectedNodeTag`. Fix was: Remove `_disableTopologyUpdate` check and always respect `SelectedNodeTag`.

Since both of these are dependent on node tag, if we disabled the topology updates we never had the cluster tag of each url, and could never run those operations according to their `SelectedNodeTag`.
Added `SingleTopologyUpdate` that fetches the node tag for all the initial urls and happens once.

- Added a check for server role = Member in `GetPreferredNodeInternal`

These changes will also fix a bug in v6.0:
https://issues.hibernatingrhinos.com/issue/RavenDB-20909/Proxy-command-loops-forever-when-the-targeted-node-is-down

### Type of change

- Bug fix

### How risky is the change?

- Moderate 

### Backward compatibility

- Changed behavior when using `_disableTopologyUpdate=true` or when targeted node is in rehab - `SelectedNodeTag` operations that previously "worked" will now throw. 

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works

### UI work

- No UI work is needed

lastav5 committed:
RavenDB-20954 - SelectedNodeTag operations reach a different node when node is in rehab or when topology updates are disabled

# Conflicts:
#	src/Raven.Client/Http/RequestExecutor.cs

lastav5 commented:
One thing that could still be a problem is when we disable the topology updates AND have a node in rehab. In that case the url that belongs to the rehab will not have its cluster tag.

karmeli87 commented on src/Raven.Client/Http/NodeSelector.cs:
```diff
@@ -111,7 +111,7 @@ private static (int Index, ServerNode Node) GetPreferredNodeInternal(NodeSelecto
             {
                 Debug.Assert(string.IsNullOrEmpty(serverNodes[i].Url) == false, $"Expected serverNodes Url not null or empty but got: \'{serverNodes[i].Url}\'");
-                if (stateFailures[i] == 0)
+                if (stateFailures[i] == 0 && serverNodes[i].ServerRole == ServerNode.Role.Member)
```
IIRC this was done for the edge case when the cluster is down, and the only live node is in rehab state... (which can't be promoted to member, because the cluster is down)

karmeli87 replied:
probably the `UnlikelyEveryoneFaultedChoice` should handle such case

lastav5 committed:
RavenDB-20954 - set server role of the nodes returned in topology update

lastav5 commented on src/Raven.Client/Http/RequestExecutor.cs:
```diff
@@ -723,6 +723,7 @@ protected async Task SingleTopologyUpdateAsync(string[] initialUrls, Guid? appli
                         await ExecuteAsync(serverNode, null, context, command, shouldRetry: false, sessionInfo: null, token: CancellationToken.None)
                             .ConfigureAwait(false);
                         serverNode.ClusterTag = command.Result.NodeTag;
+                        serverNode.ServerRole = ServerNode.Role.Member;
```
Not sure about this. Could also be promotable.. But since this is called from an executor with no topology updates, maybe it's better to have it as a member in the first place than have it stuck as promotable.

karmeli87 replied:
I think we need to take whatever the server sends us...

lastav5 replied:
It doesn't send the server role for this url. It does send the entire cluster topology but then we'll have to look up the url inside the topology and it might look different (private/public, diff format etc)

karmeli87 replied:
so why not to change the EP to send us the role as well?

ppekrol commented on src/Raven.Client/Http/RequestExecutor.cs:
```diff
@@ -697,6 +699,61 @@ private void UpdateTopologyCallback(object _)
+                catch (Exception e)
+                {
+                    serverNode.ClusterTag = "!";
+                    Logger.Info($"Error occurred while attempting to fetch the Cluster Tag for {url} in {nameof(SingleTopologyUpdateAsync)}", e);
```
IsInfoEnabled (fix the other place as well)

lastav5 replied:
other place?

lastav5 committed:
RavenDB-20954 - return server role to populate request executor's topology. CreateForFixedTopology now will not populate tags and is only for short term use

lastav5 committed:
RavenDB-20954 - address PR comments

lastav5 commented on src/Raven.Client/Http/ClusterTopology.cs:
```diff
@@ -46,6 +46,16 @@ internal void ReplaceCurrentNodeUrlWithClientRequestedUrl(string currentNodeTag,
+            if (Promotables.ContainsKey(nodeTag))
+                return ServerNode.Role.Promotable;
+
+            throw new InvalidOperationException($"The node tag {nodeTag} does not belong to the cluster");
```
should throw here or default to some role?

lastav5 committed:
RavenDB-20954 - EnsureNotPassive in GetDocumentStore, have cluster request executor for short term use without cluster tag fetching

EnsureNotPassive - So we have cluster tag for the newly created store in order to use SelectedNodeTag operations
ShortTermUse executor - Not doing SingleTopologyUpdate because this executor is for before we have stable cluster or for external requests without authorization

lastav5 committed:
RavenDB-20954 - use short term executor in more places

lastav5 committed:
RavenDB-20954 - UnlikelyEveryoneFaulted now looks first for any non-failed nodes regardless of role

lastav5 committed:
RavenDB-20954 - fix

lastav5 committed:
RavenDB-20954 - fix tabs
